controllers/machinehealthcheck_controller.go

```diff
@@ -192,3 +265,152 @@ func (r *MachineHealthCheckReconciler) clusterToMachineHealthCheck(o handler.Map
+	); err != nil {
+		r.Log.Error(err, "Unable to list MachineHealthChecks", "node", node.Name, "machine", machine.Name, "namespace", machine.Namespace)
+		return nil
+	}
+
+	var requests []reconcile.Request
+	for k := range mhcList.Items {
+		if r.hasMatchingLabels(&mhcList.Items[k], machine) {
+			key := types.NamespacedName{Namespace: mhcList.Items[k].Namespace, Name: mhcList.Items[k].Name}
+			requests = append(requests, reconcile.Request{NamespacedName: key})
+		}
+	}
+	return requests
+}
+
+func (r *MachineHealthCheckReconciler) getMachineFromNode(nodeName string) (*clusterv1.Machine, error) {
+	machineList := &clusterv1.MachineList{}
+	if err := r.Client.List(
+		context.TODO(),
+		machineList,
+		client.MatchingFields{machineNodeNameIndex: nodeName},
+	); err != nil {
+		return nil, errors.Wrap(err, "failed getting machine list")
+	}
+	if len(machineList.Items) != 1 {
+		return nil, errors.New(fmt.Sprintf("expecting one machine for node %v, got: %v", nodeName, machineList.Items))
+	}
+	return &machineList.Items[0], nil
+}
+
+func (r *MachineHealthCheckReconciler) watchClusterNodes(ctx context.Context, c client.Client, cluster *clusterv1.Cluster) error {
+	key := types.NamespacedName{Namespace: cluster.Name, Name: cluster.Name}
+	if _, ok := r.clusterNodeInformers[key]; ok {
+		// watch was already set up for this cluster
+		return nil
+	}
+
+	config, err := remote.RESTConfig(ctx, c, cluster)
+	if err != nil {
+		return errors.Wrap(err, "error fetching remote cluster config")
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "error constructing remote cluster client")
+	}
+
+	// TODO(JoelSpeed): See if we use the resync period from the manager instead of 0
+	factory := informers.NewSharedInformerFactory(k8sClient, 0)
+	nodeInformer := factory.Core().V1().Nodes().Informer()
+	go nodeInformer.Run(ctx.Done())
+
+	err = r.controller.Watch(
+		&source.Informer{Informer: nodeInformer},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(r.nodeToMachineHealthCheck)},
+	)
+	if err != nil {
+		return errors.Wrap(err, "error watching nodes on target cluster")
+	}
+
+	if r.clusterNodeInformers == nil {
+		r.clusterNodeInformers = make(map[types.NamespacedName]cache.Informer)
+	}
+
+	r.clusterNodeInformers[key] = nodeInformer
+	return nil
+}
+
+func (r *MachineHealthCheckReconciler) indexMachineByNodeName(object runtime.Object) []string {
+	machine, ok := object.(*clusterv1.Machine)
+	if !ok {
+		r.Log.Error(errors.New("incorrect type"), "expected a Machine", "type", fmt.Sprintf("%T", object))
+		return nil
+	}
+
+	if machine.Status.NodeRef != nil {
+		return []string{machine.Status.NodeRef.Name}
+	}
+
+	return nil
+}
+
+// hasMatchingLabels verifies that the MachineHealthCheck's label selector
+// matches the given Machine
+func (r *MachineHealthCheckReconciler) hasMatchingLabels(machineHealthCheck *clusterv1.MachineHealthCheck, machine *clusterv1.Machine) bool {
```

Comment:
A couple suggestions:

1. I don't think this needs to have a receiver?
1. What do you think about passing in the MHC selector & the machine labels, instead of the full objects?

Reply:
I had the receiver as I was copying what was done in the [MachineSet controller](https://github.com/kubernetes-sigs/cluster-api/blob/f4abc033a7c3036921f2108ea2b638448b454e17/controllers/machineset_controller.go#L550-L571), though I then omitted the logging from this version for some reason I can't remember 🤔 

> What do you think about passing in the MHC selector & the machine labels, instead of the full objects?

This would make it generic and reusable so sounds good to me, is [machine_helpers.go](https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machine_helpers.go) a suitable file for this generic implementation?

Reply:
+1 to machine_helpers for now, if we want to expose it in the future we can promote it

Reply:
@vincepri Ack, done this :)